devices/ICub/src/ICub.cpp

```diff
@@ -0,0 +1,385 @@
+
+#include <assert.h>
+#include <map>
+#include <mutex>
+#include <string>
+#include <vector>
+
+using namespace wearable::devices;
+
+const std::string LogPrefix = "ICub :";
+
+class WrenchPort : public yarp::os::BufferedPort<yarp::os::Bottle> {
+public:
+    std::vector<double> wrenchValues;
+    bool firstRun = true;
+    mutable std::mutex mtx;
+
+    using yarp::os::BufferedPort<yarp::os::Bottle>::onRead;
+    virtual void onRead(yarp::os::Bottle& wrench) {
+        if (!wrench.isNull()) {
+            std::lock_guard<std::mutex> lock (mtx);
+
+            this->wrenchValues.clear(); // Clear the vector of previous values
+
+            this->wrenchValues.resize(wrench.size());
+            for (size_t i = 0; i < wrench.size(); ++i) {
+                this->wrenchValues.at(i) = wrench.get(i).asDouble();
+            }
+
+            if (this->firstRun)
+            {
+                this->firstRun = false;
+            }
+        }
+        else {
+            yWarning() << LogPrefix << "[WrenchPort] read an invalid wrench bottle";
+        }
+    }
+
+    std::vector<double> wrench() {
+        std::lock_guard<std::mutex> lck (mtx);
+        return this->wrenchValues;
+    }
+
+    bool firstRunFlag() {
+        return this->firstRun;
+    }
+};
+
+class ICub::ICubImpl
+{
+public:
+    class ICubForceTorque6DSensor;
+
+    wearable::TimeStamp timeStamp;
+
+    size_t nSensors;
+    std::vector<std::string> sensorNames;
+
+    bool leftHandFTDataReceived  = false;
+    bool rightHandFTDataReceived = false;
+
+    template<typename T>
+    struct ICubDeviceSensors
+    {
+        std::shared_ptr<T> icubSensor;
+    };
+
+    std::map<std::string, ICubDeviceSensors<ICubForceTorque6DSensor>> ftSensorsMap;
+
+    const WearableName wearableName = "ICub" + wearable::Separator;
+
+    WrenchPort leftHandFTPort;
+    WrenchPort rightHandFTPort;
+};
+
+// ==========================================
+// ICub implementation of ForceTorque6DSensor
+// ==========================================
+class ICub::ICubImpl::ICubForceTorque6DSensor
+        : public wearable::sensor::IForceTorque6DSensor
+{
+public:
+    ICub::ICubImpl* icubImpl = nullptr;
+
+    // ------------------------
+    // Constructor / Destructor
+    // ------------------------
+    ICubForceTorque6DSensor(
+            ICub::ICubImpl* impl,
+            const wearable::sensor::SensorName name = {},
+            const wearable::sensor::SensorStatus status = wearable::sensor::SensorStatus::WaitingForFirstRead) //Default sensor status
+            : IForceTorque6DSensor(name, status)
+            , icubImpl(impl)
+    {
+        //Set the sensor status Ok after receiving the first data on the wrench ports
+        if (!icubImpl->leftHandFTPort.firstRunFlag() || !icubImpl->rightHandFTPort.firstRunFlag()) {
+            auto nonConstThis = const_cast<ICubForceTorque6DSensor*>(this);
+            nonConstThis->setStatus(wearable::sensor::SensorStatus::Ok);
+        }
```

Comment:
I don't understand this logic. I think this should not be here in the constructor. Let's talk f2f on how to improve the way the data is accessed from the wearable sensor and how to retrieve it using the buffered port.

My idea is that the sensors holds a reference to the `wrenchValues` buffer and (protected by a mutex) provides this data to the user. The `onRead()` callback writes new data into this buffer every time it arrives. The status should be initialized as WaitingForFirstRead in the constructor, and the `onRead()` callback updates it to `Ok`.

This approach allows to remove the hack in the wearable sensor implementation where it uses the sensor name for picking up the right data.